### ISSUE TITLE
Kotlin DSL "classpath mode" leniency doesn't capture after evaluation exceptions

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#18637]
During Gradle/Idea sync, intellij will pass the `org.gradle.kotlin.dsl.provider.mode=classpath` option to Gradle. Now, obviously any exception happening during project configuration will not fail the configuration, but will be stored in the `ClassPathModeExceptionCollector` (see `KotlinScriptEvaluator.withOptions` in the Kotlin IntelliJ Plugin).

In such cases, plugins with after evaluation hooks now see a half-configured model. For example the Kotlin Gradle Plugin runs some “post evaluate” assertions on the Project to see if it was configured correctly. This validation then fails but the thrown exception isn't captured by the `ClassPathModeExceptionCollector`, thus fails the build.

The IntelliJ Kotlin Plugin sync has been changed to not use the "classpath mode" anymore during sync as it proven to be an annoyance. This can be tracked here https://youtrack.jetbrains.com/issue/KTIJ-19823. Target is 2021.3.

The "classpath mode" will still be used by the `LoadConfigurationAction` which is responsible of loading the script context for a single script. This can be triggered by the user by clicking the corresponding button. This leniency is useful to get a good enough content assist in the editor in face of a broken build, or when editing scripts outside the build. This issue is about making that leniency better.

### Expected Behavior

When in "classpath mode" the build shouldn't fail on configuration problems, including after evaluation.

### Current Behavior

When in "classpath mode" the build will fail if an exception is thrown from an after evaluate callback.

### Potential solutions

- not running after evaluation callbacks when in classpath mode
- capturing after evaluation callbacks exceptions when in classpath mode

**Comment by donat** [comment]
🎉 🎉

**Comment by donat** [comment]
We decided to roll back the implementation and evaluate other options.

**Comment by sellmair** [comment]
@donat Can you quickly reiterate what the current situation/solution is right now?

**Comment by donat** [comment]
We did the first step and implemented the exception suppression for `afterEvaluate` callbacks. The change sparked a debate: what should happen if the one afterEvaluate callback fails, should Gradle continue configuration should it stop immediately. Continuing may drive the build to an inconsistent state as plugins expect the configuration to stop after an exception. Stopping on the other is kind of pointless considering the IntelliJ use-case. My understanding is that at least one full sync build has to be successful in IntelliJ to make the lenient mode work. So if there’s a failure later on, the IDE can capture the exception from the model building and fall back to a model loaded during the last successful sync. 

On top of that, afterEvaluate is just the tip of the iceberg in terms of callbacks. We found that other callback hooks would need different handling. This would not only complicate the codebase, but would also make it really confusing for the users and plugin authors, why their build suddenly behaves differently when in the IDE and in the command line.

Right now, our stance is to deal with the problem differently. A potential solution would be for Gradle to have support for “partial model builders”. The model building should produce a model even when the project configuration fails. The returned partial model should describe the failure message, what stage the failure happened (script evaluation, plugin application, in an afterEvaluate callback, etc). And the model that is approximately correct. For example, the Kotlin DSL script model builder could return the default script classpath even when there’s a compilation error in the settings script. This would offer a general solution for all tooling vendors to deal with the problem.
